smartsim/_core/mli/infrastructure/control/requestdispatcher.py

```diff
@@ -0,0 +1,515 @@
+        """Acquire queue lock to flush
+        :param blocking: whether to block on lock acquisition
+        :param timeout: Time to wait if blocking, before raising exception
+        """
+        return self._flush_lock.acquire(blocking=blocking, timeout=timeout)
+
+    def release(self) -> None:
+        """Release queue lock"""
+        self._flush_lock.release()
+
+    def __enter__(self) -> None:
+        """Method to use the Queue as a Context Manager"""
+        self.acquire()
+
+    def __exit__(
+        self,
+        exc_type: t.Optional[t.Type[BaseException]],
+        exc_val: t.Optional[BaseException],
+        exc_tb: t.Optional[TracebackType],
+    ) -> None:
+        """Method to release the Queue as a Context Manager"""
+        self.release()
+
+    @property
+    def model_key(self) -> str:
+        """Key of the model which needs to be run on the queued requests"""
+        return self._model_key
+
+    def put(
+        self,
+        item: InferenceRequest,
+        block: bool = False,
+        timeout: t.Optional[float] = 0.0,
+    ) -> None:
+        """Put an inference request in the queue
+        :param item: The request
+        :param block: Whether to block when trying to put the item
+        :param timeout: Time (in seconds) to wait if block==True
+        """
+        if not self.acquire(blocking=False):
+            raise Full
+        try:
+            if self.full():
+                raise Full
+            if self._first_put is None:
+                self._first_put = time.time()
+            super().put(item, block=block, timeout=timeout)
+        finally:
+            self.release()
+
+    @property
+    def _elapsed_time(self) -> float:
+        if self.empty():
+            return 0
+        return time.time() - self._first_put
+
+    @property
+    def ready(self) -> bool:
+        """True if the queue can be flushed"""
+        if self.empty():
+            return False
+        return self.full() or (self._elapsed_time >= self._batch_timeout)
+
+    def make_disposable(self) -> None:
+        """Set this queue as disposable, and never use it again after it gets flushed"""
+        self._disposable = True
+
+    @property
+    def can_be_removed(self) -> bool:
+        """Whether this queue can be deleted and garbafe collected"""
+        return self.empty() and self._disposable
+
+    def flush(self) -> list[t.Any]:
+        """Get all requests from queue
+        :return: Requests waiting to be executed
+        """
+        num_items = self.qsize()
+        self._first_put = None
+        items = []
+        for _ in range(num_items):
+            try:
+                items.append(self.get())
+            except Empty:
+                break
+
+        return items
+
+    def full(self) -> bool:
+        """Return True if the queue has reached its maximum capacity"""
+        if self._disposable:
+            return True
+        if self._batch_size <= 0:
```

Comment:
When does the batch_size get set to negative? it seems like the last if statement would be the proper comparison.

Reply:
The idea was to allow for timeout-only batching, and that by specifying a negative batch size, it would mean to only check time. But yeah, not a good solution.